core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -245,63 +302,293 @@ public static BytesReference serializeDiffClusterState(Diff diff, Version nodeVe
+        if (currentNodes.localNodeMaster() == false) {
+            ZenDiscovery.validateStateIsFromCurrentMaster(logger, currentNodes, state);
+        }
+    }
+
+    protected void handleCommitRequest(CommitClusterStateRequest request, final TransportChannel channel) {
+        ClusterState committedClusterState;
+        synchronized (lastSeenClusterStateMutex) {
+            committedClusterState = lastSeenClusterState;
+        }
+
+        // if this message somehow comes without a previous send, we won't have a cluster state
+        String lastSeenUUID = committedClusterState == null ? null : committedClusterState.stateUUID();
+        if (request.stateUUID.equals(lastSeenUUID) == false) {
+            throw new IllegalStateException("tried to commit cluster state UUID [" + request.stateUUID + "], but last seen UUID is [" + lastSeenUUID + "]");
+        }
+
+        try {
+            listener.onNewClusterState(committedClusterState, new NewClusterStateListener.NewStateProcessed() {
+                @Override
+                public void onNewClusterStateProcessed() {
+                    try {
+                        channel.sendResponse(TransportResponse.Empty.INSTANCE);
+                    } catch (Throwable e) {
+                        logger.debug("failed to send response on cluster state processed", e);
+                        onNewClusterStateFailed(e);
+                    }
                 }
-                lastSeenClusterState.status(ClusterState.ClusterStateStatus.RECEIVED);
+
+                @Override
+                public void onNewClusterStateFailed(Throwable t) {
+                    try {
+                        channel.sendResponse(t);
+                    } catch (Throwable e) {
+                        logger.debug("failed to send response on cluster state processed", e);
+                    }
+                }
+            });
+        } catch (Exception e) {
+            logger.warn("unexpected error while processing cluster state version [{}]", e, lastSeenClusterState.version());
+            throw e;
+        }
+    }
+
+    private class SendClusterStateRequestHandler implements TransportRequestHandler<BytesTransportRequest> {
+
+        @Override
+        public void messageReceived(BytesTransportRequest request, final TransportChannel channel) throws Exception {
+            handleIncomingClusterStateRequest(request, channel);
+        }
+    }
+
+    private class CommitClusterStateRequestHandler implements TransportRequestHandler<CommitClusterStateRequest> {
+        @Override
+        public void messageReceived(CommitClusterStateRequest request, final TransportChannel channel) throws Exception {
+            handleCommitRequest(request, channel);
+        }
+    }
+
+    protected static class CommitClusterStateRequest extends TransportRequest {
+
+        String stateUUID;
+
+        public CommitClusterStateRequest() {
+        }
```

Comment:
it's needed for the request to be created on the receiving node. It's used by reflection.